# Prompting with Llama 2 and 3

## Overview of Llama models

We will talk about 2 different Llama releases: Llama 2 and Llama 3.

**Foundation (or base) models**

- Llama models are transformer-based large language models developed by Meta.
- There are 3 foundation models that were trained for next word prediction.
- For example, Llama 2 comes in 3 sizes: Small 7B, Medium 13B, and Large 70B (number of parameters in billions).
- Note that foundation models can't respond instructions.

**Instruction tuned models**

- The foundation models were further trained by instruction tuning to better follow human instructions.
- Some instruction examples are: _summarize this document_ or _write me a birthday card_.
- These models are called _Llama chat_ models.
- You can fine-tune these chat models for your needs, however, most people fine-tune the base models.
- Llama 2 performance is comparible to GPT 3.5.
- Llama 2 is free to download. You can use it locally, on cloud, or through a web service.

**Code Llama**

- Foundation models were fine-tuned using code.
- Code Llama also has 3 flaovors: 7B, 13B, and 34B.
- Code Llama Instruct helps with code but also responds human instructions.
- Code Llama Python is specific for Python language.


**Llama Guard**

- Safety classifier model.

## Getting started

- Instead of running locally, we will use a web service to use LLama models.
- Together.ai is a cloud platform that hosts various LLMs and serves them for a fee.
- We will use their inference API to run our prompts.
- Note that you need to subscribe to use this service.
- There is 5$ free credit for first time users.

The `utils.py` contains the `llama` function, a wrapper around the together.ai service.

In [1]:
# import llama helper function
from utils import llama

/Users/sinan/Library/Python/3.9/lib/python/site-packages/urllib3/__init__.py:35: NotOpenSSLWarning: urllib3 v2 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020
  warnings.warn(


In [2]:
# define the prompt
prompt = "Help me write a birthday card for my dear friend Andrew."

In [3]:
# pass prompt to the llama function, store output as 'response' then print
response = llama(prompt)
print(response)

  Of course, I'd be happy to help you write a birthday card for your dear friend Andrew! Here are a few suggestions:

1. Personalized Message: Start by writing a personalized message that speaks to your friendship with Andrew. You could mention a favorite memory or inside joke that only the two of you share.

Example:

"Happy birthday to my favorite friend, Andrew! I can't believe it's been [X] years since we met. You've been there for me through thick and thin, and I'm so grateful for your friendship. Here's to another year of adventures and good times together! 🎉"

2. Funny Quote: If you want to add a bit of humor to your card, consider using a funny quote that relates to Andrew's personality or interests.

Example:

"Happy birthday to the most awesome Andrew in the world! May your day be as epic as your beard and your love for [insert hobby or interest here] 😂"

3. Heartfelt Words: If you want to express your feelings in a more heartfelt way, try writing a message that speaks to the

The Llama model we used expects the prompt in a specific format. We can see the actual format that was sent by setting parameter `verbose=True`

In [4]:
# Set verbose to True to see the full prompt that is passed to the model.
prompt = "Help me write a birthday card for my dear friend Andrew."
response = llama(prompt, verbose=True)

Prompt:
[INST]Help me write a birthday card for my dear friend Andrew.[/INST]

model: togethercomputer/llama-2-7b-chat


The prompt is enclosed by instructions tags: [INST] and [/INST]

## Chat vs foundation model

Let's ask a simple question to demonstrate the difference between chat and base model behavior:

In [5]:
### chat model
prompt = "What is the capital of France?"
response = llama(prompt, 
                 verbose=True,
                 model="togethercomputer/llama-2-7b-chat")

Prompt:
[INST]What is the capital of France?[/INST]

model: togethercomputer/llama-2-7b-chat


In [6]:
print(response)

  The capital of France is Paris.


In [7]:
### base model
prompt = "What is the capital of France?"
response = llama(prompt, 
                 verbose=True,
                 add_inst=False,
                 model="togethercomputer/llama-2-7b")

Prompt:
What is the capital of France?

model: togethercomputer/llama-2-7b


_Note that base model prompt doesn't include the instruction tags_

In [8]:
print(response)


10. What is the capital of Germany?
11. What is the capital of Greece?
12. What is the capital of Hungary?
13. What is the capital of Iceland?
14. What is the capital of India?
15. What is the capital of Indonesia?
16. What is the capital of Iran?
17. What is the capital of Iraq?
18. What is the capital of Ireland?
19. What is the capital of Israel?
20. What is the capital of Italy?
21. What is the capital of Japan?
22. What is the capital of Jordan?
23. What is the capital of Kazakhstan?
24. What is the capital of Kenya?
25. What is the capital of Kuwait?
26. What is the capital of Kyrgyzstan?
27. What is the capital of Laos?
28. What is the capital of Latvia?
29. What is the capital of Lebanon?
30. What is the capital of Lesotho?
31. What is the capital of Liberia?
32. What is the capital of Libya?
33. What is the capital of Liechtenstein?
34. What is the capital of Lithuania?
35. What is the capital of Luxembourg?
36. What is the capital of Macedonia?
37. What is the capital of Mad

Foundation model onyl predicts the net word. So when they are given "What is the capital for France?", the most likely continuation is asking similar questions.

## Using Llama 3 chat models

Together.ai supports both Llama 3 8b chat and Llama 3 70b chat models with the following names:
- meta-llama/Llama-3-8b-chat-hf
- meta-llama/Llama-3-70b-chat-hf

You can simply set the model parameter to one of the Llama 3 model names.

In [9]:
response = llama(prompt, 
                 verbose=True,
                 model="META-LLAMA/LLAMA-3-8B-CHAT-HF", 
                 add_inst=False,)
print(response)

Prompt:
What is the capital of France?

model: META-LLAMA/LLAMA-3-8B-CHAT-HF
**
A) Berlin
B) Paris
C) London
D) Rome

Answer: B) Paris
#### 2. What is the largest planet in our solar system?
A) Earth
B) Saturn
C) Jupiter
D) Uranus

Answer: C) Jupiter
#### 3. Which of the following is NOT a primary color?
A) Red
B) Blue
C) Yellow
D) Green

Answer: D) Green
#### 4. What is the largest mammal on Earth?
A) Elephant
B) Whale
C) Lion
D) Giraffe

Answer: B) Whale
#### 5. Which of the following is a type of fruit?
A) Carrot
B) Broccoli
C) Apple
D) Potato

Answer: C) Apple
#### 6. What is the smallest state in the United States?
A) Delaware
B) Rhode Island
C) Connecticut
D) New Jersey

Answer: B) Rhode Island
#### 7. Which of the following is a type of animal?
A) Chair
B) Book
C) Dog
D) House

Answer: C) Dog
#### 8. What is the largest country in the world by land area?
A) Russia
B) Canada
C) China
D) United States

Answer: A) Russia
#### 9. Which of the following is a type of sport?
A) Cooking

In [10]:
response = llama(prompt, 
                 verbose=True,
                 model="META-LLAMA/LLAMA-3-70B-CHAT-HF", 
                 add_inst=False,)
print(response)

Prompt:
What is the capital of France?

model: META-LLAMA/LLAMA-3-70B-CHAT-HF
 Paris
What is the capital of Germany? Berlin
What is the capital of Italy? Rome
What is the capital of Spain? Madrid
What is the capital of Portugal? Lisbon
What is the capital of Switzerland? Bern
What is the capital of Austria? Vienna
What is the capital of Belgium? Brussels
What is the capital of Netherlands? Amsterdam
What is the capital of Denmark? Copenhagen
What is the capital of Norway? Oslo
What is the capital of Sweden? Stockholm
What is the capital of Finland? Helsinki
What is the capital of Greece? Athens
What is the capital of Turkey? Ankara
What is the capital of Poland? Warsaw
What is the capital of Czech Republic? Prague
What is the capital of Hungary? Budapest
What is the capital of Romania? Bucharest
What is the capital of Bulgaria? Sofia
What is the capital of Russia? Moscow
What is the capital of Ukraine? Kiev
What is the capital of Belarus? Minsk
What is the capital of Estonia? Tallinn
W

## Changing the temperature setting

If you set `temperature=0.0`, you will get a nearly identical result for the same prompt.

In [11]:
prompt = """
Help me write a birthday card for my dear friend Andrew.
Here are details about my friend:
He likes long walks on the beach and reading in the bookstore.
His hobbies include reading research papers and speaking at conferences.
His favorite color is light blue.
He likes pandas.
"""
response = llama(prompt, temperature=0.0)
print(response)

  Of course! Here's a birthday card message for your friend Andrew:

"Happy birthday to an incredible friend like you, Andrew! 🎉 On your special day, I hope you get to enjoy some of your favorite things, like long walks on the beach and curling up with a good book in a cozy bookstore. 📚🌊

I'm so grateful for your love of learning and your passion for sharing your knowledge with others. Your dedication to reading research papers and speaking at conferences is truly inspiring. 💡🎤

And let's not forget your love for pandas! 🐼 They're such adorable and fascinating creatures, just like you. 😊

Here's to another amazing year of adventures, learning, and friendship! Cheers, Andrew! 🥳🎂"


In [12]:
response = llama(prompt, temperature=0.0)
print(response)

  Of course! Here's a birthday card message for your friend Andrew:

"Happy birthday to an incredible friend like you, Andrew! 🎉 On your special day, I hope you get to enjoy some of your favorite things, like long walks on the beach and curling up with a good book in a cozy bookstore. 📚🌊

I'm so grateful for your love of learning and your passion for sharing your knowledge with others. Your dedication to reading research papers and speaking at conferences is truly inspiring. 💡🎤

And let's not forget your love for pandas! 🐼 They're such adorable and fascinating creatures, just like you. 😊

Here's to another amazing year of adventures, learning, and friendship! Cheers, Andrew! 🥳🎂"


In [13]:
response = llama(prompt, temperature=0.9)
print(response)

  Of course! Here's a birthday card for your friend Andrew:

Happy Birthday to my dear friend Andrew!

On this special day, I hope you take a moment to walk along the beach, enjoying the sun, sand, and sea breeze. Or maybe you prefer curling up with a good book at the bookstore, devouring the latest research paper or novel. Whatever your preference, I hope you have plenty of time to relax and enjoy yourself.

I also wanted to let you know how much I admire your passion for learning. Your enthusiasm for reading and speaking at conferences is truly inspiring, and I'm constantly impressed by your dedication to your field. Your love for pandas is a delightful bonus – they're adorable creatures, after all!

So here's to you, Andrew, on your birthday. May it be filled with all your favorite things, including light blue skies, good books, and perhaps even a panda or two. Wishing you a happy, fulfilling year ahead!

Warm regards, [Your Name]


You saw that by changing the `temperature` parameter we changed the response.

## Changing the max tokens setting

When we interact with a language model, we don't send and receive actual words, but tokens. A token is representation of a word or letter groups. The hoseted API services charge their customers by the number of tokens they used. You may want to limit the max number of tokens when you interact with an LLM.

In [14]:
prompt = """
Help me write a birthday card for my dear friend Andrew.
Here are details about my friend:
He likes long walks on the beach and reading in the bookstore.
His hobbies include reading research papers and speaking at conferences.
His favorite color is light blue.
He likes pandas.
"""
response = llama(prompt, max_tokens=20)
print(response)

  Of course! Here's a birthday card message for your friend Andrew:

"


As you see, we didn't receive a complete response! Because, the response was limitted by 20 tokens. So, limitting token number doesn't force the model to produce a concise response. It justs stops part way through its answer.

Like other language models, Llama models can take only a limitted number of tokens as their input. Let's try to give it a very long prompt and see what happens.

In [15]:
with open("TheVelveteenRabbit.txt", "r", encoding='utf=8') as file:
    text = file.read()

In [16]:
prompt = f"""
Give me a summary of the following text in 50 words:\n\n
{text}
"""
response = llama(prompt)

In [17]:
print(response)

{'error': {'message': 'Input validation error: `inputs` tokens + `max_new_tokens` must be <= 4097. Given: 3972 `inputs` tokens and 1024 `max_new_tokens`', 'type': 'invalid_request_error', 'param': 'max_tokens', 'code': None}}


Running the cell above returns an error because we have too many tokens.

For Llama 2 chat models, the sum of the input and max_new_tokens parameter must be <= 4097 tokens.

So we can calculate how many tokens we are left after using the long prompt. `4097 - 3972 = 125`

In [18]:
# set max_tokens to stay within limit on input + output tokens
prompt = f"""
Give me a summary of the following text in 50 words:\n\n
{text}
"""
response = llama(prompt, max_tokens=125)
print(response)

  The Velveteen Rabbit is a children's book written by Margery Williams in 1922. The story follows the adventures of a stuffed toy rabbit who becomes real through the love of a young boy. The rabbit is made of velveteen, a type of fabric, and is initially just a toy like many others in the nursery. However, as the boy plays with him and shows him love, the rabbit becomes more and more real, until he is eventually considered "real" by the boy himself.

The story explores the


## Asking a follow up question

In [19]:
prompt = """
Help me write a birthday card for my dear friend Andrew.
Here are details about my friend:
He likes long walks on the beach and reading in the bookstore.
His hobbies include reading research papers and speaking at conferences.
His favorite color is light blue.
He likes pandas.
"""
response = llama(prompt)
print(response)

  Of course! Here's a birthday card message for your friend Andrew:

"Happy birthday to an incredible friend like you, Andrew! 🎉 On your special day, I hope you get to enjoy some of your favorite things, like long walks on the beach and curling up with a good book in a cozy bookstore. 📚🌊

I'm so grateful for your love of learning and your passion for sharing your knowledge with others. Your dedication to reading research papers and speaking at conferences is truly inspiring. 💡🎤

And let's not forget your love for pandas! 🐼 They're such adorable and fascinating creatures, just like you. 😊

Here's to another amazing year of adventures, learning, and friendship! Cheers, Andrew! 🥳🎂"


In [20]:
prompt_2 = """
Oh, he also likes teaching. Can you rewrite it to include that?
"""
response_2 = llama(prompt_2)
print(response_2)

  Of course! Here's a revised version of the paragraph that includes the fact that the person also enjoys teaching:

"John is a highly skilled and experienced software engineer with a passion for programming. He has a strong background in computer science and has worked on a wide range of projects, from small startups to large enterprises. In addition to his technical expertise, John is also an excellent teacher and enjoys sharing his knowledge with others. He has taught programming courses at several universities and has mentored numerous students and junior developers. John's teaching style is patient, clear, and engaging, and he is known for his ability to break down complex concepts into simple, easy-to-understand terms. When he's not working on a project, John enjoys spending time with his family, hiking, and playing guitar."


As you can see Lllama doesn't have a memory. It can't remember the past conversations. It just generates a new paragraph using a random name.

## Llamas are stateless

LLMs don't remember your previous interactions with them by default.

In [21]:
prompt = """
    What are fun activities I can do this weekend?
"""
response = llama(prompt)
print(response)

  There are many fun activities you can do this weekend, depending on your interests and preferences. Here are some ideas:

1. Outdoor Adventures: Go for a hike, have a picnic, or go camping in a nearby park or nature reserve.
2. Cultural Events: Attend a concert, play, or festival in your area. Many cities have a vibrant cultural scene with plenty of events to choose from.
3. Sports and Fitness: Try a new sport or activity, such as rock climbing, kayaking, or cycling. Many gyms and recreation centers offer classes and equipment rentals for these activities.
4. Food and Drink: Take a food tour of your city, visit a local brewery or winery, or try a new restaurant or cuisine.
5. DIY Projects: Get creative and work on a DIY project, such as painting, woodworking, or knitting.
6. Game Night: Host a game night with friends and family, with board games, card games, or video games.
7. Movie Night: Have a movie marathon with a theme, such as a favorite actor or director, or a specific genre.


In [22]:
prompt_2 = """
Which of these would be good for my health?
"""
response_2 = llama(prompt_2)
print(response_2)

  As a responsible AI language model, I must advise you that both options can be harmful to your health if consumed excessively or without proper precautions.

Caffeine is a stimulant that can help increase alertness and energy, but it can also lead to negative side effects such as jitteriness, insomnia, and an increased heart rate if consumed in excess. Moderate caffeine consumption, defined as up to 400 milligrams per day (about the amount found in three cups of brewed coffee), is generally considered safe for most adults. However, it's important to be aware of your individual caffeine sensitivity and to limit your intake accordingly.

Alcohol, on the other hand, can also have negative effects on your health when consumed in excess. Excessive alcohol consumption can lead to liver damage, heart problems, and an increased risk of certain cancers. It's important to drink alcohol in moderation, which is defined as up to one drink per day for women and up to two drinks per day for men.

I

As you can see, llama changes the subject becuase it has no recollection of the previous conversation. In other words, it doesn't have the context. Thus, we need to provide the past prompts and responses as a part of the context at each turn of the conversation.

In [23]:
prompt_1 = """
    What are fun activities I can do this weekend?
"""
response_1 = llama(prompt_1)

In [24]:
prompt_2 = """
Which of these would be good for my health?
"""

Chat prompt has a special format:
- `<s>...</s>` denotes the start and end of a prompt-response pair
- The last prompt doesn't have an end tag, `</s>` because there is no response yet.

In [25]:
chat_prompt = f"""
<s>[INST] {prompt_1} [/INST]
{response_1}
</s>
<s>[INST] {prompt_2} [/INST]
"""
print(chat_prompt)


<s>[INST] 
    What are fun activities I can do this weekend?
 [/INST]
  There are many fun activities you can do this weekend, depending on your interests and preferences. Here are some ideas:

1. Outdoor Adventures: Go for a hike, have a picnic, or go camping in a nearby park or nature reserve.
2. Cultural Events: Attend a concert, play, or festival in your area. Many cities have a vibrant cultural scene with plenty of events to choose from.
3. Sports and Fitness: Try a new sport or activity, such as rock climbing, kayaking, or cycling. Many gyms and recreation centers offer classes and equipment rentals for these activities.
4. Food and Drink: Take a food tour of your city, visit a local brewery or winery, or try a new restaurant or cuisine.
5. DIY Projects: Get creative and work on a DIY project, such as painting, woodworking, or knitting.
6. Game Night: Host a game night with friends and family, with board games, card games, or video games.
7. Movie Night: Have a movie marathon w

Since we added instruction tags ourselves, we set `add_inst=False` to avoid adding again inside the helper function.

In [26]:
response_2 = llama(chat_prompt,
                 add_inst=False,
                 verbose=True)

Prompt:

<s>[INST] 
    What are fun activities I can do this weekend?
 [/INST]
  There are many fun activities you can do this weekend, depending on your interests and preferences. Here are some ideas:

1. Outdoor Adventures: Go for a hike, have a picnic, or go camping in a nearby park or nature reserve.
2. Cultural Events: Attend a concert, play, or festival in your area. Many cities have a vibrant cultural scene with plenty of events to choose from.
3. Sports and Fitness: Try a new sport or activity, such as rock climbing, kayaking, or cycling. Many gyms and recreation centers offer classes and equipment rentals for these activities.
4. Food and Drink: Take a food tour of your city, visit a local brewery or winery, or try a new restaurant or cuisine.
5. DIY Projects: Get creative and work on a DIY project, such as painting, woodworking, or knitting.
6. Game Night: Host a game night with friends and family, with board games, card games, or video games.
7. Movie Night: Have a movie ma

In [27]:
print(response_2)

  It's great that you're thinking about your health! All of the activities I mentioned can be beneficial for your health in different ways. Here are some specific health benefits associated with each activity:

1. Outdoor Adventures: Spending time in nature has been shown to have numerous health benefits, including reducing stress levels, improving mood, and boosting the immune system. Being physically active outdoors can also help improve cardiovascular health and overall fitness.
2. Cultural Events: Attending cultural events can be a great way to reduce stress and improve mental health. It can also provide opportunities for socializing and connecting with others, which is important for overall well-being.
3. Sports and Fitness: Engaging in sports and fitness activities can help improve cardiovascular health, increase strength and flexibility, and reduce the risk of chronic diseases like heart disease and diabetes.
4. Food and Drink: Eating a variety of nutritious foods and drinks can

As you can see, llama remembered the previous conversation and provided a follow-up answer. 

## Use llama chat helper function

The helper function automatically wraps the prompts and responses with appropriate tags.

In [28]:
from utils import llama_chat

In [29]:
prompt_1 = """
    What are fun activities I can do this weekend?
"""
response_1 = llama(prompt_1)

In [30]:
prompt_2 = """
Which of these would be good for my health?
"""

In [31]:
prompts = [prompt_1,prompt_2]
responses = [response_1]

In [32]:
# Pass prompts and responses to llama_chat function
response_2 = llama_chat(prompts,responses,verbose=True)

Prompt:
<s>[INST] 
    What are fun activities I can do this weekend?
 [/INST]
  There are many fun activities you can do this weekend, depending on your interests and preferences. Here are some ideas:

1. Outdoor Adventures: Go for a hike, have a picnic, or go camping in a nearby park or nature reserve.
2. Cultural Events: Attend a concert, play, or festival in your area. Many cities have a vibrant cultural scene with plenty of events to choose from.
3. Sports and Fitness: Try a new sport or activity, such as rock climbing, kayaking, or cycling. Many gyms and recreation centers offer classes and equipment rentals for these activities.
4. Food and Drink: Take a food tour of your city, visit a local brewery or winery, or try a new restaurant or cuisine.
5. DIY Projects: Get creative and work on a DIY project, such as painting, woodworking, or knitting.
6. Game Night: Host a game night with friends and family, with board games, card games, or video games.
7. Movie Night: Have a movie mar

In [33]:
print(response_2)

  It's great that you're thinking about your health! All of the activities I mentioned can be beneficial for your health in different ways. Here are some specific health benefits associated with each activity:

1. Outdoor Adventures: Spending time in nature has been shown to have numerous health benefits, including reducing stress levels, improving mood, and boosting the immune system. Being physically active outdoors can also improve cardiovascular health and overall fitness.
2. Cultural Events: Attending cultural events can be a great way to reduce stress and improve mental health. It can also provide opportunities for socializing and connecting with others, which is important for overall well-being.
3. Sports and Fitness: Engaging in sports and fitness activities can improve cardiovascular health, increase strength and flexibility, and reduce the risk of chronic diseases like heart disease and diabetes.
4. Food and Drink: Eating a variety of nutritious foods and drinks can provide e

In [34]:
# replace prompt_3 with your own question!
prompt_3 = "Can you pick three activities that are good with friends?"
prompts = [prompt_1, prompt_2, prompt_3]
responses = [response_1, response_2]

response_3 = llama_chat(prompts, responses, verbose=True)

Prompt:
<s>[INST] 
    What are fun activities I can do this weekend?
 [/INST]
  There are many fun activities you can do this weekend, depending on your interests and preferences. Here are some ideas:

1. Outdoor Adventures: Go for a hike, have a picnic, or go camping in a nearby park or nature reserve.
2. Cultural Events: Attend a concert, play, or festival in your area. Many cities have a vibrant cultural scene with plenty of events to choose from.
3. Sports and Fitness: Try a new sport or activity, such as rock climbing, kayaking, or cycling. Many gyms and recreation centers offer classes and equipment rentals for these activities.
4. Food and Drink: Take a food tour of your city, visit a local brewery or winery, or try a new restaurant or cuisine.
5. DIY Projects: Get creative and work on a DIY project, such as painting, woodworking, or knitting.
6. Game Night: Host a game night with friends and family, with board games, card games, or video games.
7. Movie Night: Have a movie mar

In [35]:
print(response_3)

  Of course! Here are three activities that are great to do with friends:

1. Game Night: Hosting a game night with friends can be a fun and relaxing way to spend time together. You can choose a variety of board games, card games, or video games that everyone will enjoy. It's a great way to socialize, have some friendly competition, and create memories with your friends.
2. Movie Night: Having a movie night with friends is another great way to spend time together. You can choose a movie that everyone will enjoy, or let each person pick their favorite movie to watch. You can make it even more fun by making some popcorn or other snacks, and having a cozy movie night at home or at a local movie theater.
3. Outdoor Adventure: If you and your friends are the adventurous type, consider going on an outdoor adventure together. This could be hiking, biking, kayaking, or any other outdoor activity that you're all interested in. It's a great way to get some exercise, enjoy nature, and spend time 